lib/mailjet/mailer.rb

```diff
@@ -27,25 +27,187 @@ def initialize(options = {})
 class Mailjet::APIMailer
   def initialize(options = {})
-    @delivery_method_options = options.slice(
+#    send = Mailjet.Send.new
+#    if send.version.exists
+#      version = send.version
+#    else
+      @version = Mailjet.config.api_version
+#    end
+    @delivery_method_options_v3_0 = options.slice(
       :recipients, :'mj-prio', :'mj-campaign', :'mj-deduplicatecampaign',
       :'mj-templatelanguage', :'mj-templateerrorreporting', :'mj-templateerrordeliver', :'mj-templateid',
       :'mj-trackopen', :'mj-trackclick',
-      :'mj-customid', :'mj-eventpayload', :vars, :headers
+      :'mj-customid', :'mj-eventpayload', :vars, :headers,
+    )
+    @delivery_method_options_v3_1 = options.slice(
+      :'Priority', :'CustomCampaign', :'DeduplicateCampaign',
+      :'TemplateLanguage', :'TemplateErrorReporting', :'TemplateErrorDeliver', :'TemplateID',
+      :'TrackOpens', :'TrackClicks',
+      :'CustomID', :'EventPayload', :vars, :Headers,
     )
   end
 
-  def deliver!(mail)
+  def deliver!(mail, options = nil)
+#    p mail.header.fields
+    
+    if (options && options.kind_of?(Object) && options['version'].present?)
+      @version = options['version']
+    end
+    
+    if (!options.kind_of?(Object))
+      options = []
+    end
+
+    # Mailjet Send API does not support full from. Splitting the from field into two: name and email address
+    if mail[:from].nil? && Mailjet.config.default_from.present?
+      mail[:from] = Mailjet.config.default_from
+    end
+
+    if (@version == 'v3.1')
+      Mailjet::Send.create({:Messages => [setContentV3_1(mail)]})
+    else
+      Mailjet::Send.create(setContentV3_0(mail))
+    end
+  end
+
+  def setContentV3_1(mail)
+    content = {}
+    content[:TextPart] = mail.text_part.try(:decoded)
+    content[:HTMLPart] = mail.html_part.try(:decoded)
+    
+    unless mail.attachments.empty?
+      content[:Attachments] = []
+      content[:InlineAttachments] = []
+
+      mail.attachments.each do |attachment|
+        mailjet_attachment = {
+          'ContentType' => attachment.content_type.split(';')[0],
+          'Filename' => attachment.filename,
+          'Base64Content' => Base64.encode64(attachment.body.decoded)
+        }
+
+        if attachment.inline?
+          mailjet_attachment['ContentId'] = attachment.content_id
+          content[:InlineAttachments].push(mailjet_attachment)
+        else
+          content[:Attachments].push(mailjet_attachment)
+        end
+      end
+    end
+    
+    # We do the next part in purpose to accept only custom header from the user, but without accepting the mailjet related ones
+    # The mailjet related ones are stocked in another variable and directly added to the body
+    if mail.header && !mail.header.fields.empty?
+      content[:Headers] = {}
+      mail.header.fields.each do |header|
+        if header.name.start_with?('X-') && !header.name.start_with?('X-MJ') && !header.name.start_with?('X-Mailjet')
```

Comment:
if we want here keep headers starting by `X-`, but excluding strictly and only `X-MJ` and `X-Mailjet`, I would use a regex for both of these tests: [`^X-(?!\b(MJ|Mailjet)\b).*$`](http://rubular.com/r/KCBl4VXDHr):

```ruby
mail.header.fields.each do |header|
  content[:Headers][header.name] = header.value if header.name =~ /^X-(?!\b(MJ|Mailjet)\b).*$/
end
```